File: Bubble/Util/WindowListUtil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Bubble.Util
{
    internal class WindowListUtil
    {
        public class WindowInfo
        {
            public IntPtr Handle { get; private set; }

            public string Title { get; private set; }

            public string ClassName { get; private set; }

            public bool IsTopmost { get; set; }

            public WindowInfo(IntPtr handle)
            {
                Handle = handle;
                int length = Win32Util.GetWindowTextLength(Handle) * 2;
                var stringBuilder = new StringBuilder(length);
                _ = Win32Util.GetWindowText(Handle, stringBuilder, length);
                Title = stringBuilder.ToString();
                var className = new StringBuilder(256);
                _ = Win32Util.GetClassName(Handle, className, 256);
                ClassName = className.ToString();
            }
        }

        public static List<WindowInfo> GetWindowList()
        {
            var list = new List<WindowInfo>();
            Win32Util.EnumWindows((hwnd, lParam) =>
            {
                var windowInfo = new WindowInfo(hwnd);
                if (!IsCapturableWindow(windowInfo))
                {
                    return true;
                }
                list.Add(windowInfo);
                return true;
            }, IntPtr.Zero);
            return list;
        }

        private static bool IsKnownBlockedWindow(WindowInfo windowInfo)
        {
            return windowInfo.Title == "Task View" && windowInfo.ClassName == "Windows.UI.Core.CoreWindow" ||
                windowInfo.Title == "DesktopWindowXamlSource" && windowInfo.ClassName == "Windows.UI.Core.CoreWindow" ||
                windowInfo.Title == "PopupHost" && windowInfo.ClassName == "Xaml_WindowedPopupClass";
        }

        private static bool IsCapturableWindow(WindowInfo windowInfo)
        {
            if (string.IsNullOrEmpty(windowInfo.Title) ||
                windowInfo.Handle == Win32Util.GetShellWindow() ||
                !Win32Util.IsWindowVisible(windowInfo.Handle) ||
                Win32Util.GetAncestor(windowInfo.Handle, Win32Util.GA_ROOT) != windowInfo.Handle)
            {
                return false;
            }

            long style = Win32Util.GetWindowLong(windowInfo.Handle, Win32Util.GWL_STYLE);
            if ((style & Win32Util.WS_DISABLED) == Win32Util.WS_DISABLED)
            {
                return false;
            }

            style = Win32Util.GetWindowLong(windowInfo.Handle, Win32Util.GWL_EXSTYLE);
            if ((style & Win32Util.WS_EX_TOOLWINDOW) == Win32Util.WS_EX_TOOLWINDOW)
            {
                return false;
            }

            if (windowInfo.ClassName == "Windows.UI.Core.CoreWindow" || windowInfo.ClassName == "ApplicationFrameWindow")
            {
                if (Win32Util.DwmGetWindowAttribute(windowInfo.Handle, Win32Util.DWMWINDOWATTRIBUTE.DWMWA_CLOAKED, out int cloaked, sizeof(int)) == 0 &&
                    cloaked == Win32Util.DWM_CLOAKED_SHELL)
                {
                    return false;
                }
            }

            if (IsKnownBlockedWindow(windowInfo))
            {
                return false;
            }

            windowInfo.IsTopmost = (style & Win32Util.WS_EX_TOPMOST) == Win32Util.WS_EX_TOPMOST;
            return true;
        }
    }
}

File: src/Bubble/WindowListUtil.cpp
#include "pch.h"
#include "WindowListUtil.h"
#include <dwmapi.h>

bool inline MatchTitleAndClassName(std::wstring const& title, std::wstring const& className, std::wstring const& targetTitle, std::wstring const& targetClassName)
{
	return wcscmp(title.c_str(), targetTitle.c_str()) == 0 && wcscmp(className.c_str(), targetClassName.c_str()) == 0;
}

bool IsKnownBlockedWindow(std::wstring const& title, std::wstring const& className)
{
	return MatchTitleAndClassName(title, className, L"Task View", L"Windows.UI.Core.CoreWindow") || // Task View 
		MatchTitleAndClassName(title, className, L"DesktopWindowXamlSource", L"Windows.UI.Core.CoreWindow") || // XAML Islands
		MatchTitleAndClassName(title, className, L"PopupHost", L"Xaml_WindowedPopupClass"); // XAML Popups
}

bool IsCaptureableWindow(HWND const& hwnd, std::wstring const& title, std::wstring const& className, bool& isTopmost)
{
    if (title.empty() || hwnd == GetShellWindow() || !IsWindowVisible(hwnd) || GetAncestor(hwnd, GA_ROOT) != hwnd)
    {
        return false;
    }

    if (GetWindowLongW(hwnd, GWL_STYLE) & WS_DISABLED)
    {
        return false;
    }

    auto exStyle = GetWindowLongW(hwnd, GWL_EXSTYLE);
    if (exStyle & WS_EX_TOOLWINDOW)
    {
        return false;
    }

    if (wcscmp(className.c_str(), L"Windows.UI.Core.CoreWindow") == 0 || wcscmp(className.c_str(), L"ApplicationFrameWindow") == 0)
    {
        DWORD cloaked = FALSE;
        if (SUCCEEDED(DwmGetWindowAttribute(hwnd, DWMWA_CLOAKED, &cloaked, sizeof(cloaked))) && (cloaked == DWM_CLOAKED_SHELL))
        {
            return false;
        }
    }

    if (IsKnownBlockedWindow(title, className))
    {
        return false;
    }

    isTopmost = exStyle & WS_EX_TOPMOST;
    return true;
}

std::vector<Util::WindowInfo>* Util::GetWindowList()
{
	std::vector<Util::WindowInfo>* list = new std::vector<Util::WindowInfo>();
	EnumWindows([](HWND hwnd, LPARAM lParam) 
	{
		int titleLength = GetWindowTextLengthW(hwnd);
		if (titleLength > 0)
		{
			std::wstring title(++titleLength, 0);
			if (GetWindowTextW(hwnd, LPWSTR(title.data()), titleLength) == 0) return TRUE;
			int classNameLength = 256;
			std::wstring className(classNameLength, 0);
			if (GetClassNameW(hwnd, LPWSTR(className.data()), classNameLength) == 0)return TRUE;
            bool isTopmost = false;
            if (IsCaptureableWindow(hwnd, title, className, isTopmost))
            {
                std::vector<Util::WindowInfo>* list = reinterpret_cast<std::vector<Util::WindowInfo>*>(lParam);
                list->push_back(Util::WindowInfo{ hwnd,title,className,isTopmost });
            }
		}
		return TRUE;
	}, reinterpret_cast<LPARAM>(list));
	return list;
}

void Util::GetWindowList(std::vector<Util::WindowInfo> const& list)
{
    EnumWindows([](HWND hwnd, LPARAM lParam)
    {
        int titleLength = GetWindowTextLengthW(hwnd);
        if (titleLength > 0)
        {
            std::wstring title(++titleLength, 0);
            if (GetWindowTextW(hwnd, LPWSTR(title.data()), titleLength) == 0) return TRUE;
            int classNameLength = 256;
            std::wstring className(classNameLength, 0);
            if (GetClassNameW(hwnd, LPWSTR(className.data()), classNameLength) == 0)return TRUE;
            bool isTopmost = false;
            if (IsCaptureableWindow(hwnd, title, className, isTopmost))
            {
                std::vector<Util::WindowInfo>* list = reinterpret_cast<std::vector<Util::WindowInfo>*>(lParam);
                list->push_back(Util::WindowInfo{ hwnd,title,className,isTopmost });
            }
        }
        return TRUE;
    }, reinterpret_cast<LPARAM>(&list));
}
